File: ios-app/ios-app/SessionLogger.swift
import Foundation
import AVFoundation
import UIKit
import PDFKit

/// Manages session logging: transcripts with timestamps, video recording, and report generation
@MainActor
class SessionLogger: ObservableObject {
    @Published var isRecording: Bool = false
    
    // Session metadata
    private var sessionId: String
    private var sessionStartTime: Date
    
    // Transcript storage
    private var transcriptEntries: [TranscriptEntry] = []
    
    // Video recording
    private var videoWriter: AVAssetWriter?
    private var videoWriterInput: AVAssetWriterInput?
    private var pixelBufferAdaptor: AVAssetWriterInputPixelBufferAdaptor?
    private var frameCount: Int = 0
    private var videoStartTime: CMTime?
    private var lastFrameTime: CMTime?
    
    // Audio recording (for video)
    private var audioWriterInput: AVAssetWriterInput?
    
    struct TranscriptEntry: Codable {
        let timestamp: Date
        let role: String // "user" or "assistant"
        let text: String
        let isDelta: Bool // true if this is a partial update
    }
    
    struct SessionMetadata: Codable {
        let sessionId: String
        let startTime: Date
        let endTime: Date?
        let scenario: String
        let scenarioSeverity: String
        let scenarioSummary: String
        let bodyRegion: String
        let sceneObservations: [String]
    }
    
    init() {
        self.sessionId = UUID().uuidString
        self.sessionStartTime = Date()
    }
    
    // MARK: - Transcript Logging
    
    func logUserTranscript(_ text: String) {
        let entry = TranscriptEntry(
            timestamp: Date(),
            role: "user",
            text: text,
            isDelta: false
        )
        transcriptEntries.append(entry)
    }
    
    func logAgentTranscript(_ delta: String, isComplete: Bool = false) {
        if isComplete {
            // If complete, we might want to merge with last entry if it was a delta
            if let lastEntry = transcriptEntries.last,
               lastEntry.role == "assistant" && lastEntry.isDelta {
                // Replace the last delta entry with complete text
                transcriptEntries.removeLast()
                let completeText = lastEntry.text + delta
                let entry = TranscriptEntry(
                    timestamp: lastEntry.timestamp,
                    role: "assistant",
                    text: completeText,
                    isDelta: false
                )
                transcriptEntries.append(entry)
            } else {
                let entry = TranscriptEntry(
                    timestamp: Date(),
                    role: "assistant",
                    text: delta,
                    isDelta: false
                )
                transcriptEntries.append(entry)
            }
        } else {
            // Delta update - append to last entry or create new
            if let lastEntry = transcriptEntries.last,
               lastEntry.role == "assistant" && lastEntry.isDelta {
                // Update last delta entry
                transcriptEntries.removeLast()
                let updatedText = lastEntry.text + delta
                let entry = TranscriptEntry(
                    timestamp: lastEntry.timestamp,
                    role: "assistant",
                    text: updatedText,
                    isDelta: true
                )
                transcriptEntries.append(entry)
            } else {
                let entry = TranscriptEntry(
                    timestamp: Date(),
                    role: "assistant",
                    text: delta,
                    isDelta: true
                )
                transcriptEntries.append(entry)
            }
        }
    }
    
    func logSceneObservation(_ observation: String) {
        // Scene observations are logged separately but can be included in reports
    }
    
    // MARK: - Video Recording
    
    func startVideoRecording() throws {
        guard !isRecording else { return }
        
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let videoURL = documentsPath.appendingPathComponent("\(sessionId).mp4")
        
        // Remove existing file if present
        try? FileManager.default.removeItem(at: videoURL)
        
        // Create video writer
        videoWriter = try AVAssetWriter(outputURL: videoURL, fileType: .mp4)
        
        // Video settings
        let videoSettings: [String: Any] = [
            AVVideoCodecKey: AVVideoCodecType.h264,
            AVVideoWidthKey: 640,
            AVVideoHeightKey: 480,
            AVVideoCompressionPropertiesKey: [
                AVVideoAverageBitRateKey: 2000000,
                AVVideoProfileLevelKey: AVVideoProfileLevelH264HighAutoLevel
            ]
        ]
        
        videoWriterInput = AVAssetWriterInput(mediaType: .video, outputSettings: videoSettings)
        videoWriterInput?.expectsMediaDataInRealTime = true
        
        guard let videoWriterInput = videoWriterInput,
              let videoWriter = videoWriter else {
            throw NSError(domain: "SessionLogger", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to create video writer"])
        }
        
        if videoWriter.canAdd(videoWriterInput) {
            videoWriter.add(videoWriterInput)
        }
        
        // Pixel buffer adaptor for appending frames
        let sourcePixelBufferAttributes: [String: Any] = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32ARGB,
            kCVPixelBufferWidthKey as String: 640,
            kCVPixelBufferHeightKey as String: 480
        ]
        
        pixelBufferAdaptor = AVAssetWriterInputPixelBufferAdaptor(
            assetWriterInput: videoWriterInput,
            sourcePixelBufferAttributes: sourcePixelBufferAttributes
        )
        
        frameCount = 0
        videoStartTime = nil
        lastFrameTime = nil
        
        if videoWriter.startWriting() {
            isRecording = true
        } else {
            throw videoWriter.error ?? NSError(domain: "SessionLogger", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to start video writer"])
        }
    }
    
    func appendVideoFrame(_ image: UIImage) {
        guard isRecording,
              let videoWriter = videoWriter,
              let videoWriterInput = videoWriterInput,
              let pixelBufferAdaptor = pixelBufferAdaptor,
              videoWriterInput.isReadyForMoreMediaData else {
            return
        }
        
        let frameTime = CMTime(value: Int64(frameCount), timescale: 30) // 30 fps
        
        if videoStartTime == nil {
            videoStartTime = frameTime
            videoWriter.startSession(atSourceTime: frameTime)
        }
        
        guard let pixelBufferPool = pixelBufferAdaptor.pixelBufferPool else {
            return
        }
        
        var pixelBuffer: CVPixelBuffer?
        let status = CVPixelBufferPoolCreatePixelBuffer(kCFAllocatorDefault, pixelBufferPool, &pixelBuffer)
        
        guard status == kCVReturnSuccess, let buffer = pixelBuffer else {
            return
        }
        
        CVPixelBufferLockBaseAddress(buffer, [])
        defer { CVPixelBufferUnlockBaseAddress(buffer, []) }
        
        // Convert UIImage to pixel buffer
        let context = CIContext()
        if let ciImage = CIImage(image: image) {
            context.render(ciImage, to: buffer)
        }
        
        if pixelBufferAdaptor.append(buffer, withPresentationTime: frameTime) {
            frameCount += 1
            lastFrameTime = frameTime
        }
    }
    
    func stopVideoRecording() async throws -> URL {
        guard isRecording else {
            throw NSError(domain: "SessionLogger", code: 3, userInfo: [NSLocalizedDescriptionKey: "Not recording"])
        }
        
        isRecording = false
        
        guard let videoWriter = videoWriter,
              let videoWriterInput = videoWriterInput else {
            throw NSError(domain: "SessionLogger", code: 4, userInfo: [NSLocalizedDescriptionKey: "Video writer not initialized"])
        }
        
        videoWriterInput.markAsFinished()
        
        return try await withCheckedThrowingContinuation { continuation in
            videoWriter.finishWriting {
                if let error = videoWriter.error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume(returning: videoWriter.outputURL)
                }
            }
        }
    }
    
    // MARK: - PDF Export
    
    func exportTranscriptPDF(scenario: String, severity: String, summary: String, bodyRegion: String, sceneObservations: [String]) throws -> URL {
        let pdfMetaData = [
            kCGPDFContextCreator: "MedKit First-Aid Coach",
            kCGPDFContextTitle: "Session Transcript - \(sessionId)"
        ]
        
        let format = UIGraphicsPDFRendererFormat()
        format.documentInfo = pdfMetaData as [String: Any]
        
        let pageWidth = 8.5 * 72.0 // US Letter width in points
        let pageHeight = 11.0 * 72.0 // US Letter height in points
        let pageRect = CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight)
        
        let renderer = UIGraphicsPDFRenderer(bounds: pageRect, format: format)
        
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let pdfURL = documentsPath.appendingPathComponent("\(sessionId)_transcript.pdf")
        
        let data = renderer.pdfData { context in
            context.beginPage()
            
            var yPosition: CGFloat = 72.0 // Start 1 inch from top
            let margin: CGFloat = 72.0
            let contentWidth = pageWidth - (2 * margin)
            
            // Title
            let titleAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.boldSystemFont(ofSize: 24),
                .foregroundColor: UIColor.black
            ]
            let title = "Session Transcript"
            title.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: titleAttributes)
            yPosition += 40
            
            // Session info
            let infoAttributes: [NSAttributedString.Key: Any] = [
                .font: UIFont.systemFont(ofSize: 12),
                .foregroundColor: UIColor.gray
            ]
            let dateFormatter = DateFormatter()
            dateFormatter.dateStyle = .medium
            dateFormatter.timeStyle = .medium
            let sessionInfo = "Session ID: \(sessionId)\nStart Time: \(dateFormatter.string(from: sessionStartTime))"
            sessionInfo.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: infoAttributes)
            yPosition += 50
            
            // Scenario info
            if scenario != "none" {
                let scenarioAttributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.boldSystemFont(ofSize: 14),
                    .foregroundColor: UIColor.black
                ]
                let scenarioText = "Scenario: \(scenario.uppercased())\nSeverity: \(severity)\nBody Region: \(bodyRegion)\nSummary: \(summary)"
                scenarioText.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: scenarioAttributes)
                yPosition += 60
            }
            
            // Scene observations
            if !sceneObservations.isEmpty {
                let obsTitle = "Scene Observations:"
                obsTitle.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: titleAttributes)
                yPosition += 25
                
                for observation in sceneObservations {
                    let obsText = "• \(observation)"
                    let obsRect = CGRect(x: margin, y: yPosition, width: contentWidth, height: CGFloat.greatestFiniteMagnitude)
                    let obsAttributes: [NSAttributedString.Key: Any] = [
                        .font: UIFont.systemFont(ofSize: 11),
                        .foregroundColor: UIColor.darkGray
                    ]
                    let boundingRect = obsText.boundingRect(
                        with: CGSize(width: contentWidth, height: CGFloat.greatestFiniteMagnitude),
                        options: [.usesLineFragmentOrigin, .usesFontLeading],
                        attributes: obsAttributes,
                        context: nil
                    )
                    obsText.draw(in: obsRect, withAttributes: obsAttributes)
                    yPosition += boundingRect.height + 10
                    
                    if yPosition > pageHeight - 100 {
                        context.beginPage()
                        yPosition = 72.0
                    }
                }
                yPosition += 20
            }
            
            // Transcript entries
            let transcriptTitle = "Conversation Transcript:"
            transcriptTitle.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: titleAttributes)
            yPosition += 30
            
            let timeFormatter = DateFormatter()
            timeFormatter.dateStyle = .none
            timeFormatter.timeStyle = .medium
            
            for entry in transcriptEntries {
                // Check if we need a new page
                if yPosition > pageHeight - 150 {
                    context.beginPage()
                    yPosition = 72.0
                }
                
                let timeString = timeFormatter.string(from: entry.timestamp)
                let roleString = entry.role == "user" ? "User" : "Assistant"
                
                // Timestamp and role
                let headerAttributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.boldSystemFont(ofSize: 10),
                    .foregroundColor: entry.role == "user" ? UIColor.blue : UIColor.green
                ]
                let header = "[\(timeString)] \(roleString):"
                header.draw(at: CGPoint(x: margin, y: yPosition), withAttributes: headerAttributes)
                yPosition += 18
                
                // Text content
                let textRect = CGRect(x: margin + 20, y: yPosition, width: contentWidth - 20, height: CGFloat.greatestFiniteMagnitude)
                let textAttributes: [NSAttributedString.Key: Any] = [
                    .font: UIFont.systemFont(ofSize: 11),
                    .foregroundColor: UIColor.black
                ]
                let boundingRect = entry.text.boundingRect(
                    with: CGSize(width: contentWidth - 20, height: CGFloat.greatestFiniteMagnitude),
                    options: [.usesLineFragmentOrigin, .usesFontLeading],
                    attributes: textAttributes,
                    context: nil
                )
                entry.text.draw(in: textRect, withAttributes: textAttributes)
                yPosition += boundingRect.height + 20
            }
        }
        
        try data.write(to: pdfURL)
        return pdfURL
    }
    
    // MARK: - EMS Report Generation
    
    func generateEMSReport(scenario: String, severity: String, summary: String, bodyRegion: String, sceneObservations: [String]) throws -> URL {
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .full
        dateFormatter.timeStyle = .full
        
        var report = """
        ============================================
        EMS READY REPORT - FIRST AID SESSION
        ============================================
        
        SESSION INFORMATION
        --------------------
        Session ID: \(sessionId)
        Start Time: \(dateFormatter.string(from: sessionStartTime))
        End Time: \(dateFormatter.string(from: Date()))
        Duration: \(String(format: "%.1f", Date().timeIntervalSince(sessionStartTime))) seconds
        
        """
        
        if scenario != "none" {
            report += """
            SCENARIO DETAILS
            ----------------
            Type: \(scenario.uppercased())
            Severity: \(severity.uppercased())
            Body Region: \(bodyRegion)
            Summary: \(summary)
            
            """
        }
        
        if !sceneObservations.isEmpty {
            report += """
            SCENE OBSERVATIONS
            -----------------
            """
            for (index, observation) in sceneObservations.enumerated() {
                report += "\(index + 1). \(observation)\n"
            }
            report += "\n"
        }
        
        report += """
        CONVERSATION TRANSCRIPT
        -----------------------
        """
        
        let timeFormatter = DateFormatter()
        timeFormatter.dateStyle = .none
        timeFormatter.timeStyle = .medium
        
        for entry in transcriptEntries {
            let timeString = timeFormatter.string(from: entry.timestamp)
            let roleString = entry.role == "user" ? "USER" : "ASSISTANT"
            report += "[\(timeString)] \(roleString): \(entry.text)\n\n"
        }
        
        report += """
        
        KEY INFORMATION SUMMARY
        ------------------------
        """
        
        // Extract key information
        var userStatements: [String] = []
        var assistantInstructions: [String] = []
        
        for entry in transcriptEntries {
            if entry.role == "user" {
                userStatements.append(entry.text)
            } else {
                assistantInstructions.append(entry.text)
            }
        }
        
        report += """
        User Statements (\(userStatements.count) total):
        """
        for statement in userStatements {
            report += "\n• \(statement)"
        }
        
        report += """
        
        
        Assistant Instructions (\(assistantInstructions.count) total):
        """
        for instruction in assistantInstructions {
            report += "\n• \(instruction)"
        }
        
        report += """
        
        
        ============================================
        END OF REPORT
        ============================================
        """
        
        // Save as text file
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let reportURL = documentsPath.appendingPathComponent("\(sessionId)_EMS_Report.txt")
        try report.write(to: reportURL, atomically: true, encoding: .utf8)
        
        return reportURL
    }
    
    // MARK: - Session Management
    
    func reset() {
        sessionId = UUID().uuidString
        sessionStartTime = Date()
        transcriptEntries.removeAll()
        frameCount = 0
        videoStartTime = nil
        lastFrameTime = nil
    }
    
    func getSessionId() -> String {
        return sessionId
    }
}

File: backend/session_logger.py
"""
Backend session logger: tracks transcripts, scenarios, and generates reports.
"""

import json
import os
import time
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Optional


class TranscriptEntry:
    def __init__(self, timestamp: float, role: str, text: str):
        self.timestamp = timestamp
        self.role = role  # "user" or "assistant"
        self.text = text

    def to_dict(self) -> dict:
        return {
            "timestamp": self.timestamp,
            "datetime": datetime.fromtimestamp(self.timestamp).isoformat(),
            "role": self.role,
            "text": self.text,
        }


class SessionLogger:
    def __init__(self, session_id: str):
        self.session_id = session_id
        self.start_time = datetime.now()
        self.transcript_entries: List[TranscriptEntry] = []
        self.scene_observations: List[Dict] = []
        self.scenario_updates: List[Dict] = []
        self.tool_calls: List[Dict] = []
        
        # Current scenario state
        self.current_scenario = "none"
        self.current_severity = "minor"
        self.current_summary = ""
        self.current_body_region = ""

    def log_user_transcript(self, text: str):
        """Log a user transcript entry."""
        entry = TranscriptEntry(time.time(), "user", text)
        self.transcript_entries.append(entry)
        print(f"[SessionLogger] User: {text[:50]}...")

    def log_assistant_transcript(self, text: str, is_delta: bool = False):
        """Log an assistant transcript entry."""
        if is_delta and self.transcript_entries and self.transcript_entries[-1].role == "assistant":
            # Append to last entry if it's a delta
            self.transcript_entries[-1].text += text
        else:
            entry = TranscriptEntry(time.time(), "assistant", text)
            self.transcript_entries.append(entry)
        if not is_delta:
            print(f"[SessionLogger] Assistant: {text[:50]}...")

    def log_scene_observation(self, observation: str):
        """Log a scene observation."""
        entry = {
            "timestamp": time.time(),
            "datetime": datetime.now().isoformat(),
            "observation": observation,
        }
        self.scene_observations.append(entry)

    def log_scenario_update(self, scenario: str, severity: str, summary: str, body_region: str):
        """Log a scenario update."""
        self.current_scenario = scenario
        self.current_severity = severity
        self.current_summary = summary
        self.current_body_region = body_region
        
        entry = {
            "timestamp": time.time(),
            "datetime": datetime.now().isoformat(),
            "scenario": scenario,
            "severity": severity,
            "summary": summary,
            "body_region": body_region,
        }
        self.scenario_updates.append(entry)

    def log_tool_call(self, tool_name: str, params: dict):
        """Log a tool call."""
        entry = {
            "timestamp": time.time(),
            "datetime": datetime.now().isoformat(),
            "tool": tool_name,
            "params": params,
        }
        self.tool_calls.append(entry)

    def save_session_log(self, output_dir: Optional[str] = None) -> str:
        """Save session log as JSON file."""
        if output_dir is None:
            output_dir = os.path.join(os.getcwd(), "session_logs")
        
        Path(output_dir).mkdir(parents=True, exist_ok=True)
        
        log_data = {
            "session_id": self.session_id,
            "start_time": self.start_time.isoformat(),
            "end_time": datetime.now().isoformat(),
            "duration_seconds": (datetime.now() - self.start_time).total_seconds(),
            "transcript_entries": [e.to_dict() for e in self.transcript_entries],
            "scene_observations": self.scene_observations,
            "scenario_updates": self.scenario_updates,
            "tool_calls": self.tool_calls,
            "current_scenario": {
                "scenario": self.current_scenario,
                "severity": self.current_severity,
                "summary": self.current_summary,
                "body_region": self.current_body_region,
            },
        }
        
        filename = f"{self.session_id}_session_log.json"
        filepath = os.path.join(output_dir, filename)
        
        with open(filepath, "w") as f:
            json.dump(log_data, f, indent=2)
        
        print(f"[SessionLogger] Saved session log to {filepath}")
        return filepath

    def generate_ems_report(self, output_dir: Optional[str] = None) -> str:
        """Generate EMS-ready text report."""
        if output_dir is None:
            output_dir = os.path.join(os.getcwd(), "session_logs")
        
        Path(output_dir).mkdir(parents=True, exist_ok=True)
        
        report_lines = [
            "=" * 50,
            "EMS READY REPORT - FIRST AID SESSION",
            "=" * 50,
            "",
            "SESSION INFORMATION",
            "-" * 50,
            f"Session ID: {self.session_id}",
            f"Start Time: {self.start_time.strftime('%Y-%m-%d %H:%M:%S')}",
            f"End Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
            f"Duration: {(datetime.now() - self.start_time).total_seconds():.1f} seconds",
            "",
        ]
        
        if self.current_scenario != "none":
            report_lines.extend([
                "SCENARIO DETAILS",
                "-" * 50,
                f"Type: {self.current_scenario.upper()}",
                f"Severity: {self.current_severity.upper()}",
                f"Body Region: {self.current_body_region}",
                f"Summary: {self.current_summary}",
                "",
            ])
        
        if self.scene_observations:
            report_lines.extend([
                "SCENE OBSERVATIONS",
                "-" * 50,
            ])
            for i, obs in enumerate(self.scene_observations, 1):
                dt = datetime.fromisoformat(obs["datetime"])
                report_lines.append(f"{i}. [{dt.strftime('%H:%M:%S')}] {obs['observation']}")
            report_lines.append("")
        
        report_lines.extend([
            "CONVERSATION TRANSCRIPT",
            "-" * 50,
        ])
        
        for entry in self.transcript_entries:
            dt = datetime.fromtimestamp(entry.timestamp)
            role = entry.role.upper()
            report_lines.append(f"[{dt.strftime('%H:%M:%S')}] {role}: {entry.text}")
            report_lines.append("")
        
        report_lines.extend([
            "",
            "KEY INFORMATION SUMMARY",
            "-" * 50,
        ])
        
        user_statements = [e.text for e in self.transcript_entries if e.role == "user"]
        assistant_instructions = [e.text for e in self.transcript_entries if e.role == "assistant"]
        
        report_lines.append(f"User Statements ({len(user_statements)} total):")
        for statement in user_statements:
            report_lines.append(f"  • {statement}")
        
        report_lines.append("")
        report_lines.append(f"Assistant Instructions ({len(assistant_instructions)} total):")
        for instruction in assistant_instructions:
            report_lines.append(f"  • {instruction}")
        
        if self.tool_calls:
            report_lines.extend([
                "",
                "TOOL CALLS",
                "-" * 50,
            ])
            for call in self.tool_calls:
                dt = datetime.fromisoformat(call["datetime"])
                report_lines.append(f"[{dt.strftime('%H:%M:%S')}] {call['tool']}: {json.dumps(call['params'])}")
        
        report_lines.extend([
            "",
            "=" * 50,
            "END OF REPORT",
            "=" * 50,
        ])
        
        report_text = "\n".join(report_lines)
        
        filename = f"{self.session_id}_EMS_Report.txt"
        filepath = os.path.join(output_dir, filename)
        
        with open(filepath, "w") as f:
            f.write(report_text)
        
        print(f"[SessionLogger] Generated EMS report: {filepath}")
        return filepath
